### ISSUE TITLE
PIN verification fails on IAS-SEC card

### ISSUE THREAD

**Issue by qsantos** [OpenSC/OpenSC#1240]
My initial bug report was erroneous, so here is a better description of the problem.

### Problem Description

OpenSC fails to verify the PIN on several IAS-SEC smartcards.

### Proposed Resolution

Per @redsectors investigation, it seems like the card is expecting a PIN padded to a length of 24 bytes using zeros. The following patch is sufficient to fix the verify command for this card:

```diff
diff --git a/src/libopensc/card-iasecc.c b/src/libopensc/card-iasecc.c
index a364a394..07754e64 100644
--- a/src/libopensc/card-iasecc.c
+++ b/src/libopensc/card-iasecc.c
@@ -1847,6 +1847,13 @@ iasecc_chv_verify(struct sc_card *card, struct sc_pin_cmd_data *pin_cmd,
                apdu.data = pin_cmd->pin1.data;
                apdu.datalen = pin_cmd->pin1.len;
                apdu.lc = pin_cmd->pin1.len;
+
+               unsigned char buf[24];
+               memset(buf, 0, 24);
+               memcpy(buf, pin_cmd->pin1.data, pin_cmd->pin1.len);
+               apdu.data = buf;
+               apdu.datalen = 24;
+               apdu.lc = 24;
        }
        else if ((card->reader->capabilities & SC_READER_CAP_PIN_PAD) && !pin_cmd->pin1.data && !pin_cmd->pin1.len)   {
                rv = iasecc_chv_verify_pinpad(card, pin_cmd, tries_left);
```

However, this patch may break functionality for other cards, and does not address the issue in other commands (i.e. changing the PIN). It seems to me that the best course of actions would be to introduce a workaround in `iasecc_pin_cmd()`, before dispatching, which would detect the affected cards and pad both `pin1` and `pin2` accordingly.

### Steps to reproduce

Introduce the card in reader, run `pkcs15-tool --verify`, and enter correct PIN.

### Logs

```$ OPENSC_DEBUG=2 pkcs15-tool --verify
Using reader with a card: Alcor Micro AU9560 00 00
0x7ff21a328340 13:07:53.480 [pkcs15-tool] card.c:200:sc_connect_card: called
Please enter PIN [PIN_Global]: 
0x7ff21a328340 13:07:55.972 [pkcs15-tool] sec.c:217:sc_pin_cmd: returning with: -1214 (PIN code or key incorrect)
Operation failed: PIN code or key incorrect
0x7ff21a328340 13:07:55.976 [pkcs15-tool] ctx.c:898:sc_release_context: called
```

[With OPENSC_DEBUG=9](https://gist.github.com/qsantos/53f1e6adcade74f531c2de186bd6e7c0).

**Comment by martinpaljak** [comment]
The problem is (usually) not in the reader, but the card. Are you sure that you have a "IAS/ECC Morpho MinInt - Agent Card" ?

**Comment by qsantos** [comment]
Yes, I was still a bit confused between the roles of the reader and the card itself. Apparently, the card is issued by the Impremerie Nationale, which issue cards for several institutions, and probably for the MinInt. I am not sure how card drivers work however. Is there such a thing as a card model number?

I realize that this may not be the most appropriate place to ask these question, but thanks for your help!

**Comment by dengert** [comment]
Could be a bad card.  Where did you get this card?
Does it work with any other software, reader or other computer?

Can you try this command:
 `opensc-tool -c default -f`
this will use the default driver and try and list any files on the card. 


If you don't not expect this to be a   'IAS/ECC Morpho MinInt - Agent Card' 
you can try and edit the opensc.conf  and replace the drivers = line with a line with the drivers you think might work. 

It may contain a different application that that will respond to a SELECT AID in another driver.



In the log, OpenSC has tried to determine what type of card you have so it can select a driver. 
The first part tries to match the ATR and it looks like the ATR matches:
`card-iasecc.c:352:iasecc_match_card: 'IAS/ECC Morpho MinInt - Agent Card' card matched`
but then the next line is
`card-iasecc.c:630:iasecc_init: called`
and the first thing the card driver tries to do is a SELECT AID command Which is a special case of a SELECT FILE command
`00 A4 04 00 06 4D 49 4F 4D 43 54` 

But the AID is `4D 49 4F 4D 43 54`  thit the first 4 bits being a `4' ISO-7816-4 Section 8.2.1.2 Application identifier Table 90 says:
"'0' to '9' - Reserved for backward compatibility with ISO/IEC 7812-1[3] (see annex D)"
So this may not be a true AID, but an Issuer identification number. 

So this is an ambiguous case Thus the card may not be treating this as a SELECT AID, because it starts with `4`.  

So this could be:
- a problem with your specific card. 
- This type of card (based on ATA) is being used for some other application and the card-iasecc.c has assumed it  must be  'IAS/ECC Morpho MinInt - Agent Card` but in reality it is not. 

The ATR `3b:df:96:00:80:31:fe:45:00:31:b8:64:04:1f:ec:c1:73:94:01:80:82:90:00:ec`  can be parsed by
https://smartcard-atr.appspot.com/ 
This says `- Application selection: by full DF name`  which means it supports SELECT AID and should not return `6A 86` "Incorrect parameters P1-P2" if this was a true AID. 

In any case card-iasecc.c should be looked at closer and handle the failure differently, and it should be doing the SELECT FILE `00 A4 04 00 06 4D 49 4F 4D 43 54` in the match part, and if it fails assume it is not a 'IAS/ECC Morpho MinInt - Agent Card'

**Comment by frankmorgner** [comment]
@qsantos please check if https://github.com/OpenSC/OpenSC/pull/1241 fixes your issue.

**Comment by qsantos** [comment]
@dengert, thanks for the information. I have ran `opensc-tool -c DRIVER -f` for each of the drivers supported by OpenSC with the following command:

```
for driver in $(opensc-tool --list-driver | tail -n+2 | cut -d' ' -f3); do echo $driver; opensc-tool -c $driver -f; done
```

Most returned `Incorrect parameters in APDU`, but two were interesting:

```
$ opensc-tool -c sc-hsm -f
Using reader with a card: Alcor Micro AU9560 00 00
3f00 type: DF, size: 0
select[N/A] lock[N/A] delete[N/A] create[N/A] rehab[N/A] inval[N/A] list[N/A] 

Empty directory
```

```
$ opensc-tool -c jpki -f
Using reader with a card: Alcor Micro AU9560 00 00
3f00 type: DF, size: 0
select[NONE] lock[NEVR] delete[NEVR] create[NEVR] rehab[N/A] inval[N/A] list[NONE] 

sc_list_files() failed: Not supported
```

If I understand correctly, the third output line lists the available commands. But then, I am not sure how to interpret the various states. Or maybe it represent some settings on the card?

Anyway, an empty directory seems plausible to me, seems the PKI has not yet been transferred to the smart card (first checking whether it was usable under Linux). Thus, `sc-hsm` looks like the more promising driver.

How correct are my conclusions? Are there more commands I can run? More generally, I am not sure where to look for information about using smart cards; some pointers would be greatly appreciated!

Anyway, thanks for your patience!

**Comment by frankmorgner** [comment]
The IAS/ECC contains an obvious problem with the applet selection. I fixed this in https://github.com/frankmorgner/OpenSC/commit/892e8925e330e120a5a6dd5e232cd7c2c9156da9. @qsantos, could you use this change and try again with your card?

**Comment by dengert** [comment]
Try @frankmorgner change. It could be a newer version of the IAS/ECC card application is doing more tests of the command then previous versions that accepted  an obviously bad command. 

I doubt that the sc_hsm (driver for CardContact devices) or jpki (Japanese PKI card) would work with your card. I was asking for you to only try the default driver. 

You still did not say where you got the card or if you expected it to be an IAS/ECC card. 
What I was hinting at is some other card distributor is using the same card with the same ATR but putting a different applet on the card. Or your card is defective.

**Comment by qsantos** [comment]
@dengert sorry, it was 3f00, not jpki (which seems more likely); I did run the test for the default driver, but figured out that there was a small probability that trying other drivers might give some clue

I do not know much about smart cards model, so I though that "IAS/ECC Morpho MinInt - Agent Card" was specific to the French Ministry of the Interior, hence my reply about the Imprimerie Nationale. However, it seems that it is indeed a IAS/ECC (see below); I am still a bit confused about the meaning of the full name.

@frankmorgner @dengert I realized I was actually not using the most up-to-date version during my initial tests: `zsh` has some weird behaviors when updating binaries in `$PATH`, and was still using `/usr/bin/opensc-tool` instead of `/usr/local/bin/opensc-tool` (i.e., I failed to `rehash`).

I actually obtain the following output with the latest version (with, or without frankmorgner/OpenSC@892e892).

```
$ opensc-tool -n
Using reader with a card: Alcor Micro AU9560 00 00
IAS-ECC
$ opensc-tool -f                                                                                                          
Using reader with a card: Alcor Micro AU9560 00 00
3f00 type: DF, size: 0
select[N/A] lock[N/A] delete[N/A] create[N/A] rehab[N/A] inval[N/A] list[N/A] 

sc_list_files() failed: Not supported
```

Sorry for the waste of time using the wrong version of OpenSC. I still have no idea how to interpret the output however, is there some reading I can do for this purpose?

**Comment by qsantos** [comment]
As an addition, I have been able to list the PINs and certificates of the smart card using `pkcs15-tool`. However, neither Firefox and `pkcs15-tool` manage to bind with the card, as if the PIN were incorrect (the number of tries is also decremented on the card). Would providing debug output when trying to bind with the card be useful? If so, what command would work best to that effect?

**Comment by frankmorgner** [comment]
Yes, a debug log would be helpful. run pkcs11-tool in gdb and get a backtrace of the segfault, if possible.

**Comment by qsantos** [comment]
I fixed `pkcs11-tool`'s segfault. It was due to having both the repository and the git versions installed: for some reason, `pkcs11-tool` ignores `/usr/local/lib` and used the incorrect shared object file. Removing the repository version and making a clean install of the git version lets `pkcs11-tool` work correctly:

```$ pkcs11-tool -L    
Available slots:
Slot 0 (0x0): Alcor Micro AU9560 00 00
  token label        : PIN_Global (pr_XXXXXXXX_151-002-
  token manufacturer : Imprimerie Nationale
  token model        : PKCS#15
  token flags        : login required, rng, token initialized, PIN initialized, user PIN locked
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : 610001498632519A
  pin min/max        : 4/12
```

However, this did not fix the PIN issue (it did say `final user PIN try` before) and, as you can see in the above log, I will now need to have it unlocked before running further tests. When I will have done so, do you think the debug output of `pkcs15-tool --verify-pin` would be useful?

**Comment by redsectors** [comment]
Hello,

looks like you own same card as myself :)
Please note that this card does not work of the box with openSC...
Indeed PIN code sent to this card must be 24 bytes length whatever the length of your PIN so padding is needed.
I have a dirty patch that totally break other ias-ecc cards so i do not publish it here :)

See here:

https://github.com/OpenSC/OpenSC/issues/1096

BR

Elie

**Comment by frankmorgner** [comment]
@redsectors / @qsantos , the PIN length is determined in [`iasecc_pin_get_policy()`](https://github.com/OpenSC/OpenSC/blob/baa709ff74631336694c609ccfc2b7c32bd3cf53/src/libopensc/card-iasecc.c#L2210-L2221). The *correct (TM)* fix would be finding out why your pin length is incorrect and maybe write a specific workaround for your card.

**Comment by redsectors** [comment]
I detected this behaviour while snooping traffic sent to my smart card.

I ended up with below dirty patch... To be honest i am not enough skilled to do the same with the function you describe ^^




         static int iasecc_chv_verify(struct sc_card *card, struct sc_pin_cmd_data *pin_cmd, int *tries_left)
       {
        struct sc_context *ctx = card->ctx;
         struct sc_acl_entry acl = pin_cmd->pin1.acls[IASECC_ACLS_CHV_VERIFY];
         struct sc_apdu apdu;
         int rv;
         unsigned char pinwithpadding[24];

        LOG_FUNC_CALLED(ctx);
        sc_log(ctx, "Verify CHV PIN(ref:%i,len:%i,acl:%X:%X)", pin_cmd->pin_reference, pin_cmd->pin1.len,
                        acl.method, acl.key_ref);

        if (acl.method & IASECC_SCB_METHOD_SM)   {
                rv = iasecc_sm_pin_verify(card, acl.key_ref, pin_cmd, tries_left);
                LOG_FUNC_RETURN(ctx, rv);
        }

        if (pin_cmd->pin1.data && !pin_cmd->pin1.len)   {
                sc_format_apdu(card, &apdu, SC_APDU_CASE_1, 0x20, 0, pin_cmd->pin_reference);
        }
        else if (pin_cmd->pin1.data && pin_cmd->pin1.len)   {
                sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x20, 0, pin_cmd->pin_reference);

                /* Patch to force pin lenght to 24 bytes with zero padding */
                /*
                apdu.data = pin_cmd->pin1.data;
                apdu.datalen = pin_cmd->pin1.len;
                apdu.lc = pin_cmd->pin1.len;
                */
                sc_log(ctx, "Forcing pin length for Imprimerie Nationale agent card (%i)", card->type);
                memset(pinwithpadding, 0, sizeof(pinwithpadding));
                memcpy(pinwithpadding, pin_cmd->pin1.data, pin_cmd->pin1.len);
                apdu.data = pinwithpadding;
                apdu.datalen = 24;
                apdu.lc = 24;
                /* End patch */
`

**Comment by qsantos** [comment]
Nice investigation @redsectors, you went much farther than I did! Thanks for the information @frankmorgner. So, I have been exploring `card-iassec.c` a bit, trying to find the appropriate way to include a patch for our cards. For the convenience of other people willing to play around in the code, I give below what seems to be the relevant sequence of events:

First, `iasecc_pin_verify()` calls `iasecc_pin_get_policy()` (where PIN length is determined), *then* sets `pin_cmd.pin1.data = data` (`data` is an argument of the function) and later `iasecc_chv_verify()` (where the patch currently resides). `iasecc_chv_verify()` prepares the APDU by setting `apdu.data = pin_cmd->pin1.data` and then sends it by calling `sc_transmit_apdu()`. `sc_transmit_apdu()` (`apdu.c`) just pass `apdu.data` which goes unchanged at least until to the smart card reader driver (as a dynamic library through `dlopen`).

In other words, the value of the APDU is not padded between the common API (`iassec_pin_cmd()` which calls `iassec_pin_verify()`) and the eventual transmission to the hardware. Additionally, `iasecc_pin_get_policy()` is called before `pin_cmd.pin1.data` is set and can therefore not effect it. However, wherever we put the patch for pin1, we would still need to address pin2, which is used by `iassec_pin_change()` and `iassec_pin_reset()`.

So to keep changes clean and minimal, it seems to me that the best approach would be to treat the problem as early as possible, i.e. in `iasecc_pin_cmd()`, by having a call to a new function, that would detect the offending cards, and unilaterally pads pin1 and pin2 (regardless of `data->cmd`).

However, it remains to find a way to identify the card without false positive (which introduce a regression). In #1096, @redsectors provided an AID, but this does not seem to match with my own card, whose ATR I reproduce below (no AID is detected by [gpg.jar](https://github.com/OpenSC/OpenSC/wiki/JavaCards), if I even understand what an AID is).

```
3b:df:96:00:80:31:fe:45:00:31:b8:64:04:1f:ec:c1:73:94:01:80:82:90:00:ec
```

**Comment by frankmorgner** [comment]
@redsectors will you have a look into a cleaner patch with the approach suggested from @qsantos ?

**Comment by frankmorgner** [comment]
@qsantos your initial log shows that a [PIN of max 24/min 4 is detected](https://gist.github.com/qsantos/53f1e6adcade74f531c2de186bd6e7c0#file-log-txt-L2447).

There is some code for padding the PIN to the maximum. Strangely, however, this is done with the padding character `0xFF` instead of `0x00` (as done with the patch above). The [IAS/ECC specification](https://eevertti.vrk.fi/documents/2634109/2858578/GIXEL+244+IAS+ECC+v1+0+1UK.pdf) doesn't mention PIN padding, so it seems to be a vendor specific decision. If so, we need to uniquely identify your vendor somehow. I fear, this can only be done by someone with greater insights. @viktorTarasov ?

**Comment by viktorTarasov** [comment]
The alternative could be two commit here below:
https://github.com/viktorTarasov/OpenSC-SM/commit/3a925819990296187e0385b70dcf1d01de8312ee and https://github.com/viktorTarasov/OpenSC-SM/commit/b575594eceb72a33bb08934762feeddc0729bff7

PIN padding is not in the IAS-ECC specification, but it's legitimate from the PKCS#15 point of view.

PIN padding is traditionally needed for the Oberthur's cards.
PINs are padded with 0xFF utill the length 64.
The pad length and pad character are correctly described in the PKCS#15 meta-data, so there is no need to fix them 'in a hard manner' somewhere in the card driver.

Now the question could be: should the PINs be padded in the PKCS#15 framework and delivered to the card driver already expanded, or it's up to card driver to expand the PIN ?
In the commits above it's the card driver who do this work.

**Comment by muktupavels** [comment]
> Now the question could be: should the PINs be padded in the PKCS#15 framework and delivered to the card driver already expanded, or it's up to card driver to expand the PIN ?

I am confused how that is supposed to work... `opensc.h` has `filled in by the card driver` comment for `pad_length`. But it looks like data is already filled before driver!?

How about introducing helper function to expand pin? Something like this:
```
diff --git a/src/libopensc/card-iasecc.c b/src/libopensc/card-iasecc.c
index ea3d99b4..45eec103 100644
--- a/src/libopensc/card-iasecc.c
+++ b/src/libopensc/card-iasecc.c
@@ -2345,13 +2345,37 @@ iasecc_keyset_change(struct sc_card *card, struct sc_pin_cmd_data *data, int *tr
 }
 
 
+static int
+pad_pin_if_needed(unsigned int flags, struct sc_pin_cmd_pin pin, unsigned char *pin_data, size_t *pin_len)
+{
+	if ((flags & SC_PIN_CMD_NEED_PADDING) != 0) {
+		if (pin.pad_length > *pin_len)
+			return 0;
+
+		memset(pin_data, pin.pad_char, pin.pad_length);
+		memcpy(pin_data, pin.data, pin.len);
+		*pin_len = pin.pad_length;
+	}
+	else {
+		if ((size_t) pin.len > *pin_len)
+			return 0;
+
+		memcpy(pin_data, pin.data, pin.len);
+		*pin_len = pin.len;
+	}
+
+	return 1;
+}
+
+
 static int
 iasecc_pin_change(struct sc_card *card, struct sc_pin_cmd_data *data, int *tries_left)
 {
 	struct sc_context *ctx = card->ctx;
 	struct sc_apdu apdu;
 	unsigned reference = data->pin_reference;
-	unsigned char pin_data[0x100];
+	unsigned char pin_data[0x100], pin1_data[0x80], pin2_data[0x80];
+	size_t pin1_len = sizeof(pin1_data), pin2_len = sizeof(pin2_data);
 	int rv;
 
 	LOG_FUNC_CALLED(ctx);
@@ -2365,27 +2389,27 @@ iasecc_pin_change(struct sc_card *card, struct sc_pin_cmd_data *data, int *tries
 		}
 	}
 
-	if (!data->pin1.data && data->pin1.len)
+	if ((!data->pin1.data && data->pin1.len) || !pad_pin_if_needed(data->flags, data->pin1, pin1_data, &pin1_len))
 		LOG_TEST_RET(ctx, SC_ERROR_INVALID_ARGUMENTS, "Invalid PIN1 arguments");
 
-	if (!data->pin2.data && data->pin2.len)
+	if ((!data->pin2.data && data->pin2.len) || !pad_pin_if_needed(data->flags, data->pin2, pin2_data, &pin2_len))
 		LOG_TEST_RET(ctx, SC_ERROR_INVALID_ARGUMENTS, "Invalid PIN2 arguments");
 
-	rv = iasecc_pin_verify(card, data->pin_type, reference, data->pin1.data, data->pin1.len, tries_left);
+	rv = iasecc_pin_verify(card, data->pin_type, reference, pin1_data, pin1_len, tries_left);
 	sc_log(ctx, "iasecc_pin_cmd(SC_PIN_CMD_CHANGE) pin_verify returned %i", rv);
 	LOG_TEST_RET(ctx, rv, "PIN verification error");
 
-	if ((unsigned)(data->pin1.len + data->pin2.len) > sizeof(pin_data))
+	if (pin1_len + pin2_len > sizeof(pin_data))
 		LOG_TEST_RET(ctx, SC_ERROR_BUFFER_TOO_SMALL, "Buffer too small for the 'Change PIN' data");
 
 	if (data->pin1.data)
-		memcpy(pin_data, data->pin1.data, data->pin1.len);
+		memcpy(pin_data, pin1_data, pin1_len);
 	if (data->pin2.data)
-		memcpy(pin_data + data->pin1.len, data->pin2.data, data->pin2.len);
+		memcpy(pin_data + pin1_len, pin2_data, pin2_len);
 
 	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x24, 0, reference);
 	apdu.data = pin_data;
-	apdu.datalen = data->pin1.len + data->pin2.len;
+	apdu.datalen = pin1_len + pin2_len;
 	apdu.lc = apdu.datalen;
 
 	rv = sc_transmit_apdu(card, &apdu);
@@ -2513,6 +2537,8 @@ static int
 iasecc_pin_cmd(struct sc_card *card, struct sc_pin_cmd_data *data, int *tries_left)
 {
 	struct sc_context *ctx = card->ctx;
+	unsigned char pin_data[256];
+	size_t pin_len = sizeof(pin_data);
 	int rv;
 
 	LOG_FUNC_CALLED(ctx);
@@ -2522,7 +2548,8 @@ iasecc_pin_cmd(struct sc_card *card, struct sc_pin_cmd_data *data, int *tries_le
 
 	switch (data->cmd)   {
 	case SC_PIN_CMD_VERIFY:
-		rv = iasecc_pin_verify(card, data->pin_type, data->pin_reference, data->pin1.data, data->pin1.len, tries_left);
+		pad_pin_if_needed(data->flags, data->pin1, pin_data, &pin_len);
+		rv = iasecc_pin_verify(card, data->pin_type, data->pin_reference, pin_data, pin_len, tries_left);
 		break;
 	case SC_PIN_CMD_CHANGE:
 		if (data->pin_type == SC_AC_AUT)
```

Padding probably is needed also in reset/unblock case... But I can not test this part as I get segfault when doing `C_Login` with `CKU_SO` followed by `C_InitPIN`. Segfaults in `sc_pkcs15init_create_pin` calling `profile->ops->create_pin`. :(

**Comment by muktupavels** [comment]
@viktorTarasov do you plan to create pull request?

**Comment by misterzed88** [comment]
Min/max PIN lengths can be defined both at the IAS-ECC card level and at the PKSC #15 level. Some cards signal padding requirements by settings the same min/max PIN length at the IAS-ECC card level.

As you have already suggested, the IAS-ECC card driver should consider the SC_PIN_CMD_NEED_PADDING flag and the pad_char field. It should ideally heuristically deduce padding requirements by also taking the card level info into account. Padding must be correctly implemented for all functions (verify, change and unblock) and also for the PIN-pad case. This can get a bit hairy.

I have done all of this work in an updated IAS-ECC driver, soon to be published.

**Comment by misterzed88** [comment]
The following PR contains changes for solving PIN padding requirements, for all use cases (verify, change and unblock) and also for PIN-pad entry:
https://github.com/OpenSC/OpenSC/pull/2070

**Comment by Jakuje** [comment]
If I read it right, this was fixed with #2070 so I will close this issue.
If it is still a problem for somebody, please, open a new issue describing your problems.
